src/base/utils/string.cpp

```diff
@@ -45,110 +45,103 @@ namespace
-                    else
-                        return leftChar < rightChar;
+                if ((posL == left.size()) || (posR == right.size()))
+                    return (left.size() - right.size());  // when a shorter string is another string's prefix, shorter string place before longer string
 
+                const QChar leftChar = (m_caseSensitivity == Qt::CaseSensitive) ? left[posL] : left[posL].toLower();
+                const QChar rightChar = (m_caseSensitivity == Qt::CaseSensitive) ? right[posR] : right[posR].toLower();
+                if (leftChar == rightChar) {
+                    // compare next character
                     ++posL;
                     ++posR;
                 }
-
-                int startL = posL;
-                while ((posL < left.size()) && left[posL].isDigit())
-                    ++posL;
-                int numL = left.midRef(startL, posL - startL).toInt();
-
-                int startR = posR;
-                while ((posR < right.size()) && right[posR].isDigit())
-                    ++posR;
-                int numR = right.midRef(startR, posR - startR).toInt();
-
-                if (numL != numR)
-                    return (numL < numR);
-
-                // Strings + digits do match and we haven't hit string end
-                // Do another round
+                else if (leftChar.isDigit() && rightChar.isDigit()) {
+                    // Both are digits, compare the numbers
+                    const auto consumeNumber = [](const QString &str, int &pos) -> int
+                    {
+                        const int start = pos;
+                        while ((pos < str.size()) && str[pos].isDigit())
+                            ++pos;
+                        return str.midRef(start, (pos - start)).toInt();
+                    };
+
+                    const int numL = consumeNumber(left, posL);
+                    const int numR = consumeNumber(right, posR);
+                    if (numL != numR)
+                        return (numL - numR);
+
+                    // String + digits do match and we haven't hit the end of both strings
+                    // then continue to consume the remainings
+                }
+                else {
+                    return (leftChar.unicode() - rightChar.unicode());
+                }
             }
-            return false;
         }
 
-    private:
         QCollator m_collator;
-        const bool m_caseSensitive;
+        const Qt::CaseSensitivity m_caseSensitivity;
     };
 }
 
-bool Utils::String::naturalCompareCaseSensitive(const QString &left, const QString &right)
+int Utils::String::naturalCompare(const QString &left, const QString &right, const Qt::CaseSensitivity caseSensitivity)
 {
     // provide a single `NaturalCompare` instance for easy use
     // https://doc.qt.io/qt-5/threads-reentrancy.html
+    if (caseSensitivity == Qt::CaseSensitive) {
```

Comment:
`if-else` isn't needed anymore. You can pass param into NaturalCompare constructor. It was my initial idea of using Qt::CaseSensitivity in it.

Reply:
>if-else isn't needed anymore. 

Won't that require `template <Qt::CaseSensitivity> class NaturalCompare` ?
Otherwise case sensitivity in this function is fixed on first usage.

Reply:
Oops... I forgot about static variables.
Then it would be better to have caseSensitivity as template param

Reply:
The second idea is we can reuse QCollator of NaturalCompare with different case sensitivity option:
1. Turn NaturalCompare into NaturalComparator (or NaturalCollator) since we don't use it as functor now
2. Remove param from its constructor
3. Remove operator()
4. Implement public compare(Qt::CaseSensitivity) and use it (it uses QCollator internally or custom code as operator() now)
5. naturalCompare can have only one comparator object (per thread) and call its compare() method passing caseSensitivity param to it

Reply:
>Implement public compare(Qt::CaseSensitivity) and use it (it uses QCollator internally or custom code as operator() now)

I already finished it.
Yet deciding Qt::CaseSensitivity at runtime (need a call to QCollator::setCaseSensitivity()) might induce performance hit: https://github.com/qt/qtbase/blob/3fe74b76fd0eaf39d4c6681e2edca5adbf107883/src/corelib/tools/qcollator.cpp#L195
Are you ok with it?

Reply:
1. ...
2. ...
3. ...
4. ...
5. ...
6. Then it turns out that the NaturalCompare class is not needed. All needed logic can be implemented directly in naturalCompare function (having thread local QCollator object if needed).

Reply:
>Yet deciding Qt::CaseSensitivity at runtime (need a call to QCollator::setCaseSensitivity()) might induce performance hit

We don't have shared copies of QCollator so detach shouldn't affect performance.

Reply:
>We don't have shared copies of QCollator so detach shouldn't affect performance.

Also, `d->dirty = true;` is causing some trouble, it causing init() here: https://github.com/qt/qtbase/blob/3fe74b76fd0eaf39d4c6681e2edca5adbf107883/src/corelib/tools/qcollator_macx.cpp#L94
This won't be a issue now when Qt::CaseSensitivity is rarely changed, but might be a hidden issue later if we change it rapidly.

Reply:
>This isn't a issue now when Qt::CaseSensitivity is rarely changed, but might be a hidden issue later if Qt::CaseSensitivity is changing rapidly.

You still can choose template-based approach.

Reply:
>You still can choose template-based approach.

In short, that will require naturalCompare & NaturalCompare to become templates and move to header file.
IMO exposing NaturalCompare via header is a bit too much, I think I just stay with the current PR state.